src/analyses/variable-sensitivity/abstract_enviroment.cpp

```diff
@@ -0,0 +1,585 @@
+/*******************************************************************\
+
+ Module: analyses variable-sensitivity
+
+ Author: Thomas Kiley, thomas.kiley@diffblue.com
+
+\*******************************************************************/
+#include "abstract_enviroment.h"
+#include <functional>
+#include <stack>
+#include <map>
+#include <analyses/variable-sensitivity/abstract_object.h>
+#include <analyses/variable-sensitivity/constant_abstract_value.h>
+#include <analyses/variable-sensitivity/struct_abstract_object.h>
+#include <analyses/variable-sensitivity/pointer_abstract_object.h>
+#include <analyses/variable-sensitivity/array_abstract_object.h>
+#include <analyses/ai.h>
+#include <util/simplify_expr.h>
+
+
+#ifdef DEBUG
+#include <iostream>
+#endif
+
+/*******************************************************************\
+
+Function: abstract_environmentt::eval
+
+  Inputs:
+   expr - the expression to evaluate
+
+ Outputs: The abstract_object representing the value of the expression
+
+ Purpose: Evaluate the value of an expression
+
+\*******************************************************************/
+
+abstract_object_pointert abstract_environmentt::eval(
+  const exprt &expr, const namespacet &ns) const
+{
+  assert(!is_bottom);
+  typedef std::function<abstract_object_pointert(const exprt &)> eval_handlert;
+  std::map<irep_idt, eval_handlert> handlers=
```

Comment:
I do accept that the code below is correct, and maybe it's even beautiful. But it's probably quite a bit longer than a sequence of 5 `if`/`else if` statements, isn't it? The overhead of constructing the map might also outweigh the performance gains of doing log-N lookups instead of going through the cascade of ifs.